File: user_study/solutions/q1_ans.c
// Correct answer: CWE-125: Out-of-bounds Read
// Source: From cwe-snippets, ./snippets_1/non-compliant/C/0025.c



int getValueFromArray(int *array, int len, int index) {
    int value;
    
    // check that the array index is less than the maximum
    // length of the array
    if (index < len && index >= 0) {
        // get the value at the specified index of the array
        value = array[index];
    }
    
    // if array index is invalid then output error message
    // and return value indicating error
    else {
        printf("Invalid array index\n");
        value = -1;
    }
    
    return value;
}

File: user_study/solutions/q1_ans.java
// Correct answer: CWE-190: Integer Overflow or Wraparound
// Source: from cwe-snippets, snippets_90/non-compliant/Java/cwe-0190/byte_max_multiply_01.java
/*
 * The Byte.MAX_VALUE in Java is 127, and when we attempt to multiply 127 by 2, the expected mathematical result is 254. 
 * However, due to the byte data type in Java being an 8-bit signed integer with a range from -128 to 127, 
 * this operation will lead to an overflow. When the result of the multiplication (254) is cast back to a byte, 
 * it will not fit within the byte range. 
 */


// Solution:
// The solution is to check if the multiplication by 2 would cause an overflow before performing the operation.
public void func() throws Throwable {
    byte data = Byte.MAX_VALUE;

    if (data > 0) {
        // Check if multiplication by 2 would cause overflow.
        if (data > Byte.MAX_VALUE / 2) {
            // Handle the potential overflow case.
            System.out.println("Warning: Operation would cause overflow.");
        } else {
            // Safe to perform the operation.
            byte result = (byte)(data * 2);
            IO.writeLine("result: " + result);
        }
    }
}
